fix(components): Fix state warning errors

zerostatic opened the pull request:
Closes #1243

A few components are giving the following error in the console:
```
Did not properly initialize state during construction. 
Expected state to be an object, but it was undefined.
```

#### Changelog

The following components were fixed:
AccordionItem
ComposedModal
ContentSwitcher
FileUploader
NumberInput
OverflowMenu
PaginationV2
ProgressIndicator
RadioButtonGroup
SearchLayoutButton
TimePicker.js
Tooltip.js

#### Notes

A few components were checking is the state variable was defined. I had to change this now to check if state is an empty object. This is the instances were you see this: 
```
let emptyState =
      Object.keys(state).length === 0 && state.constructor === Object;
```

![giphy](https://media.giphy.com/media/pkqqFEa3Gqjfy/giphy.gif)

liviudobrea commented:
There's also this https://github.com/facebook/react/issues/12523
Personally, I would rather have it explicitly defined, so I wouldn't rely on gDFSP setting up initial state for me.
I don't expect React to add a fallback for when state is null on first instance and gDFSP, rather they'll probably suppress the error at some point with a warning. But that doesn't fix the initial issue.
That being said, I'm with @zerostatic on this one.

asudoh commented:
Weighing in - My earlier comment is not about coding preference, but about not breaking existing code. This PR need more work in this regard. Thanks!

asudoh commented:
Rewind - Looking at `gDSFP()` logic again I see a breakage won’t happen. My apologies for the false alarm. @zerostatic Do you want to update `gDSFP()` logic to remove the obsolete code handling undefined `state`? If not that’s fine and I can do that afterwards. Thanks!

asudoh commented:
Rewind - Looking at `gDSFP()` logic again I see a breakage won’t happen. My apologies for the false alarm. @zerostatic Do you want to update `gDSFP()` logic to remove the obsolete code handling undefined state? If not that’s fine and I can do that afterwards. Thanks!

zerostatic commented:
@asudoh I'm not sure I understand, 
Now that I've added an empty state object: `state = {}`
I replaced the code that checks for undefined state with code that checks if state is an empty object, like:

I replaced this:
```if (!state) {...```

with this:
``` if (Object.keys(state).length === 0 && state.constructor === Object) {....```

Is there a cleaner way to handle this?

asudoh commented:
@zerostatic Thank you for asking and my apologies for the confusion. Let’s take your change to `AccordionItem.js` as an example - Given you removed “undefined `state`” condition, we can remove the logic handling that condition in `gDSFP()`, like:

```diff
--- AccordionItem.js.orig	2018-09-03 06:36:01.000000000 +0900
+++ AccordionItem.js	2018-09-03 06:36:13.000000000 +0900
@@ -51,8 +51,8 @@
   };
 
   static getDerivedStateFromProps({ open }, state) {
-    const { prevOpen } = state || {};
-    return state && prevOpen === open
+    const { prevOpen } = state;
+    return prevOpen === open
       ? null
       : {
           open,
```

Don’t hesitate to ping me if you have any further questions - Thanks!

zerostatic commented:
@asudoh Thanks for the explanation! I've made the updates. Let me know if I can do anything to improve it.

asudoh commented:
Thanks @zerostatic for your updates! Would you want to make similar changes to `gDSFP()` in files where you added empty initial states? If you don’t feel like it it’ll be no problem - Let us know. Thanks!

zerostatic commented:
@asudoh No problem...the other gDSFP's have been updated.

zerostatic commented on src/components/NumberInput/NumberInput.js:
```diff
@@ -5,6 +5,8 @@ import Icon from '../Icon';
 import classNames from 'classnames';
 
 export default class NumberInput extends Component {
+  state = { value: '' };
```
Sure thing...this has been updated. Thanks!
